File: 0213-house-robber-ii/0213-house-robber-ii.py
class Solution:
    #tabulation
    def solve(self,nums,dp,ind):

        prev=nums[0]
        prev2=0
        for i in range(1,len(nums)):
            pick=nums[i]+prev2
            skip=prev
            curr=max(pick,skip)
            prev2=prev
            prev=curr
        return prev    
            


    def rob(self, nums: List[int]) -> int:
        n=len(nums)

        if n==1:
            return nums[0]

        arr1=nums[0:n-1]
        dp1=[-1]*len(arr1) 

        arr2=nums[1:n]
        dp2=[-1]*len(arr2)

        ans1=self.solve(arr1,dp1,len(arr1)-1)   
        ans2=self.solve(arr2,dp2,len(arr2)-1)

    

        return max(ans1,ans2)

                 
        

File: 0213-house-robber-ii/0213-house-robber-ii.java
class Solution {
    public int rob(int[] nums) {
        int n=nums.length;
        if (n==1){
            return nums[0];
        }
        int first=solve(nums,0,n-2);
        int last=solve(nums,1,n-1);
        return Math.max(first,last);
        
    }
    public int solve(int[] nums, int start, int end){
        int[] dp=new int[end-start+1];
        dp[0]=nums[start];
        if(dp.length>1){
            dp[1]=Math.max(nums[start],nums[start+1]);
        }
        for(int i=2;i<dp.length;i+=1){
            int rob=nums[start+i]+dp[i-2];
            int notrob=dp[i-1];
            dp[i]=Math.max(rob,notrob);
        }
        return dp[dp.length-1];
    }
}